### PR TITLE
NV-1010 - Create webhook parser for SES (email)

### PR DESCRIPTION
**Pull request by gitstart**
### What kind of change does this PR introduce?

Implemented getMessageId and parseEventBody to parse SES(email) webhook.

### Why was this change needed?

Fixes https://github.com/novuhq/novu/issues/1446

### Other information:

Example [JSON Response](https://docs.aws.amazon.com/ses/latest/dg/event-publishing-retrieving-firehose-examples.html)

```
{
  "eventType": "Delivery",
  "mail": {
    "timestamp": "2016-10-19T23:20:52.240Z",
    "source": "sender@example.com",
    "sourceArn": "arn:aws:ses:us-east-1:123456789012:identity/sender@example.com",
    "sendingAccountId": "123456789012",
    "messageId": "EXAMPLE7c191be45-e9aedb9a-02f9-4d12-a87d-dd0099a07f8a-000000",
    "destination": [
      "recipient@example.com"
    ],
    "headersTruncated": false,
    "headers": [
      {
        "name": "From",
        "value": "sender@example.com"
      },
      {
        "name": "To",
        "value": "recipient@example.com"
      },
      {
        "name": "Subject",
        "value": "Message sent from Amazon SES"
      },
      {
        "name": "MIME-Version",
        "value": "1.0"
      },
      {
        "name": "Content-Type",
        "value": "text/html; charset=UTF-8"
      },
      {
        "name": "Content-Transfer-Encoding",
        "value": "7bit"
      }
    ],
    "commonHeaders": {
      "from": [
        "sender@example.com"
      ],
      "to": [
        "recipient@example.com"
      ],
      "messageId": "EXAMPLE7c191be45-e9aedb9a-02f9-4d12-a87d-dd0099a07f8a-000000",
      "subject": "Message sent from Amazon SES"
    },
    "tags": {
      "ses:configuration-set": [
        "ConfigSet"
      ],
      "ses:source-ip": [
        "192.0.2.0"
      ],
      "ses:from-domain": [
        "example.com"
      ],
      "ses:caller-identity": [
        "ses_user"
      ],
      "ses:outgoing-ip": [
        "192.0.2.0"
      ],
      "myCustomTag1": [
        "myCustomTagValue1"
      ],
      "myCustomTag2": [
        "myCustomTagValue2"
      ]      
    }
  },
  "delivery": {
    "timestamp": "2016-10-19T23:21:04.133Z",
    "processingTimeMillis": 11893,
    "recipients": [
      "recipient@example.com"
    ],
    "smtpResponse": "250 2.6.0 Message received",
    "reportingMTA": "mta.example.com"
  }
}
```

[Event types](https://docs.aws.amazon.com/ses/latest/dg/event-publishing-retrieving-firehose-contents.html)

image

---
This code was written and reviewed by GitStart Community. Growing great engineers, one PR at a time.

### CODE REVIEW

**Review by p-fernandez — CHANGES_REQUESTED** [review]

File: providers/ses/src/lib/ses.provider.ts [hunk 1]
```diff
@@ -85,6 +87,63 @@ export class SESEmailProvider implements IEmailProvider {
...
+        return EmailEventStatusEnum.REJECTED;
+      case 'Open':
+        return EmailEventStatusEnum.OPENED;
+      case 'Click':
+        return EmailEventStatusEnum.CLICKED;
+      case 'DeliveryDelay':
+        return EmailEventStatusEnum.DELAYED;
+      case 'Subscription':
+        return EmailEventStatusEnum.UNSUBSCRIBED;
+    }
```
[the remark this answers is not in the document]
Comment (gitstart) [reply providers/ses/src/lib/ses.provider.ts:144]: @p-fernandez how do you suggest that this be handled?

[the remark this answers is not in the document]
Comment (p-fernandez) [reply providers/ses/src/lib/ses.provider.ts:144]: Considering the documentation (https://user-images.githubusercontent.com/32432927/226950725-dce29ec1-6fbe-4eea-b66a-a2e4377742f1.png), that the status is not an action or outcome of the event and more a state of the subscriber resource I'd skip it from the switch but leaving a comment in the mapping, as the code would handle to return `undefined` in `parseEventBody`.
@scopsy what do you think about this?

[the remark this answers is not in the document]
Comment (gitstart) [reply providers/ses/src/lib/ses.provider.ts:144]: @scopsy Gentle reminder about this :) Think we should just skip it as recommended by @p-fernandez ?

[the remark this answers is not in the document]
Comment (gitstart) [reply providers/ses/src/lib/ses.provider.ts:144]: Alright then. Will push an update for this asap.

File: providers/ses/src/lib/ses.provider.ts [hunk 2]
```diff
@@ -85,6 +87,62 @@ export class SESEmailProvider implements IEmailProvider {
...
+        return EmailEventStatusEnum.SENT;
+      case 'Reject':
+        return EmailEventStatusEnum.REJECTED;
+      case 'Open':
+        return EmailEventStatusEnum.OPENED;
+      case 'Click':
+        return EmailEventStatusEnum.CLICKED;
+      case 'DeliveryDelay':
+        return EmailEventStatusEnum.DELAYED;
+      // case 'Subscription':
```
Comment (p-fernandez) [inline providers/ses/src/lib/ses.provider.ts:142]: Just remove it. If you want to leave a function comment before line 124 mentioning that `Subscription` event status is not considered for the reasons explained would be a bit more meaningful.

**Review by p-fernandez — APPROVED** [review]
🌟

**Pull request merged**